### PR TITLE
Created FlattenedContiguousArrayIterator for FA iteration

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
Created iterator which has nextSlot() method to sequentially
return slots found inside a Flattened Array.
-Implements getIndex(), setIndex(int) getter and setter method for
index within the flattened array.
-Implements save(ObjectIteratorState) and restore(ObjectIteratorState)
methods to save scan and end pointers to objectIteratorState, or set
the scan and end pointers from the objectIteratorState.
-Implements getObject() to return pointer to the array.

Created new CopmactSchemeFixupObject method: fixupFlattenedArrayObject,
which allows the gc compacter to copmact flattened arrays using
FlattenedContiguousArrayIterator.

Added ValueTypeTest for flattenedArray compact, using -Xcompactgc
flag.

Signed-off-by: Oussama Saoudi <oussama.saoudi@ibm.com>

### CODE REVIEW

**Comment by OussamaSaoudi** [comment]
@dmitripivkine

File: runtime/gc_structs/FlattenedContiguousArrayIterator.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+	MMINLINE void initialize(J9Object *objectPtr)
+	{
+		MM_GCExtensionsBase *extensions = MM_GCExtensionsBase::getExtensions(_omrVM);
+		GC_ArrayObjectModel *arrayObjectModel = &(extensions->indexableObjectModel);
+		J9Class *clazzPtr = J9GC_J9OBJECT_CLAZZ(objectPtr, extensions);
+
+		_elementStride = J9ARRAYCLASS_GET_STRIDE(clazzPtr);
+		_basePtr = (fomrobject_t *)arrayObjectModel->getDataPointerForContiguous(_arrayPtr);
+		_scanPtr = (_basePtr);
```
Comment (dmitripivkine) [inline runtime/gc_structs/FlattenedContiguousArrayIterator.hpp:79]: please remove brackets

File: runtime/gc_structs/FlattenedContiguousArrayIterator.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+class GC_FlattenedContiguousArrayIterator
+{
+private:
+	J9IndexableObject *_arrayPtr;					/**< pointer to the array object being iterated */
+	GC_MixedObjectIterator _mixedObjectIterator;	/**< Object iterator which iterates over field of each element */
+	uintptr_t _elementStride;						/**< Size of each element, including padding */
+	fj9object_t *_basePtr;							/**< pointer to the first array slot, where element 0 is */
+	fj9object_t *_scanPtr;							/**< pointer to the current array element's first slot */
+	fj9object_t *_endPtr;							/**< pointer to the array slot where the iteration will terminate */
+	J9Class *_elementClass;							/**< Pointer to class of the elements in the flattened array */
```
Comment (dmitripivkine) [inline runtime/gc_structs/FlattenedContiguousArrayIterator.hpp:51]: please do not try to align comments, just keep single space after `;`

File: runtime/gc_structs/FlattenedContiguousArrayIterator.hpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ */
+
+#if !defined(FLATTENEDCONTIGUOUSARRAYITERATOR_HPP_)
+#define FLATTENEDCONTIGUOUSARRAYITERATOR_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+
+#include "ArrayObjectModel.hpp"
+#include "Bits.hpp"
```
Comment (dmitripivkine) [inline runtime/gc_structs/FlattenedContiguousArrayIterator.hpp:35]: do we need this to be included?

File: runtime/gc_structs/FlattenedContiguousArrayIterator.hpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		objectIteratorState->_endPtr  = _endPtr;
+	}
+
+	MMINLINE J9Object *getObject()
+	{
+		return (J9Object *)_arrayPtr;
+	}
+
+	GC_FlattenedContiguousArrayIterator(OMR_VM *omrVM, J9Object *objectPtr)
+		: _arrayPtr((omrarrayptr_t)objectPtr)
```
Comment (dmitripivkine) [inline runtime/gc_structs/FlattenedContiguousArrayIterator.hpp:131]: The declared type of `_arrayPtr` is `J9IndexableObject *`, so I believe it is better to cast to this type

**Review by amicic — APPROVED** [review]

**Comment by dmitripivkine** [comment]
@tajila Would you please review test update part?

**Review by dmitripivkine — APPROVED** [review]

**Comment by dmitripivkine** [comment]
Jenkins test sanity xlinuxvalst jdknext

**Comment by dmitripivkine** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Review by tajila — APPROVED** [review]

**Pull request merged**